app/controllers/dois_controller.rb

```diff
@@ -0,0 +1,52 @@
+# frozen_string_literal: true
+
+class DoisController < ApplicationController
+  class DoiIsMintingError < RuntimeError; end
+
+  def create
+    resource = find_resource
+
+    authorize(resource, :edit?)
+    raise DoiIsMintingError if doi_present_or_minting?(resource)
+
+    MintDoiAsync.call(resource)
+  rescue DoiIsMintingError
+    # I don't think we need to do anything special here, just skip over the call
+    # to MintDoiAsync and redirect back where we came from?
+  ensure
+    redirect_to resource_path(params[:resource_id])
+  end
+
+  def show
+    @resource = find_resource
+    render :show, layout: false
+  end
+
+  private
+
+    # @note We have some ridiculous hoops to jump through here.
+    # 1. We can only send an actual, undecorated ActiveRecord through ActiveJob
+    # 2. The ResourceDecorator and associated subclasses define a method called
+    #    `resource_with_doi` because the WorkVersion cannot currently receive a
+    #    doi, only the parent work can.
+    # Therefore, we need to decorate the resource, ask it what the
+    # resource-with-doi is, then un-decorate the returned object.
+    def find_resource
+      # Note the param is called :resource_id, but it's actually a uuid
+      undecorated_resource = FindResource.call(params[:resource_id])
+      decorated_resource = ResourceDecorator.decorate(undecorated_resource)
+      resource_with_doi = decorated_resource.resource_with_doi
+
+      # If resource_with_doi is decorated (`__getobj__`), undecorate before
+      # returning
+      if resource_with_doi.respond_to?(:__getobj__)
+        resource_with_doi.__getobj__
+      else
+        resource_with_doi
+      end
```

Comment:
I may be missing something, but couldn't this be simplified to:

``` ruby
resource = FindResource.call(params[:resource_id])

return resource.try(:work) || resource
```

It just returns whatever the work is. Are we dealing with collections too?

Reply:
We _totally could be_ dealing with collections in the future, yes. Also there's no guarantee that the decorator has not overloaded `#work` to return a WorkDecorator. `__getobj__` is a feature of `SimpleDelegator` (the thing we use to make decorators) that gives you access to the raw, undecorated object being wrapped

Reply:
But, FindResource is returning the undecordated object, which is ultimately what you want to return, correct? You've defined `resource_with_doi` to be used to display a MintableDoiComponent but here it seems like you're jumping through a lot of hoops just to reuse that method. I understand that reasoning, but for clarity, wouldn't 

``` ruby
return resource.work if resource.is_a?(WorkVersion)
resource
```

Or something to that effect, be easier to understand? I realize that probably violates some principle because the concept of what does nor does not have a doi is now in two different places.

Reply:
Yes, then you've duplicated the logic of "a WorkVersion shouldn't have a DOI, only its parent work should"

Reply:
Another option might be to put `#resource_with_doi` onto the models themselves, not on the decorator, since that is a fairly important piece of business logic?

Reply:
Yes, I was just going to suggest that. Or have `WorkVersion#doi` return nil. On the other hand, work versions will _eventually_ have dois so all this business can go away.

Reply:
OK, I'll relocate `#resource_with_doi` into the models, I think that's a better place for it anyways than hiding out in a decorator